src/signature.c

```diff
@@ -46,115 +46,301 @@
 
 #warning "TODO pick signing scheme"
 #if defined(SWUPD_LINUX_ROOTFS)
-#define IMPL IMPL_FAKE
+#define IMPL IMPL_REAL
 #endif
 
 #if IMPL != IMPL_FAKE
 
-static X509_STORE *create_store(const char *, const char *, const char *);
 
-static char *VERIF_FAIL = "Signature verification failed";
-static char *XSTORE_FAIL = "XSTORE creation failed";
+bool signing_enabled = false;
+
+static bool validate_signature(FILE *, FILE *);
+static bool validate_certificate(void);
+static int verify_callback(int, X509_STORE_CTX *);
+static bool get_pubkey(const char *);
+static bool get_certificates(void);
 
 static bool initialized = false;
+static EVP_PKEY *pkey;
+static X509 *cert;
+static char *chain;
+static char *crl;
+static char *certificate;
+static char *ca_dirname;
 
-static X509_STORE *x509_store = NULL;
 
 bool signature_initialize(const char *ca_cert_filename)
 {
 	if (initialized) {
 		return true;
 	}
+
 	OpenSSL_add_all_algorithms();
 	ERR_load_crypto_strings();
-	x509_store = create_store(ca_cert_filename, NULL, NULL);
-	if (x509_store == NULL) {
+
+	if (!get_certificates()) {
+		return false;
+	}
+
+	if (!validate_certificate()) {
 		ERR_free_strings(); // undoes ERR_load_crypto_strings
 		EVP_cleanup();      // undoes OpenSSL_add_all_algorithms
 		return false || (IMPL == IMPL_FORGIVE);
 	}
+
+	if (!get_pubkey(ca_cert_filename)) {
+		return false || (IMPL == IMPL_FORGIVE);
+	}
+
 	initialized = true;
 	return true;
 }
 
 void signature_terminate(void)
 {
 	if (initialized) {
-		X509_STORE_free(x509_store); // undocumented...
 		ERR_free_strings();	  // undoes ERR_load_crypto_strings
-		EVP_cleanup();		     // undoes OpenSSL_add_all_algorithms
+		EVP_cleanup();		  // undoes OpenSSL_add_all_algorithms
+		EVP_PKEY_free(pkey);
+		X509_free(cert);
 		initialized = false;
 	}
 }
 
+bool get_certificates(void)
+{
+	/* TODO - get paths and certificate names from Linux environment */
+
+
+	return true;
+}
+
 bool signature_verify(const char *data_filename, const char *sig_filename)
 {
-	BIO *bio_data = NULL;
-	BIO *bio_sig = NULL;
-	PKCS7 *pkcs7 = NULL;
-	int ret;
+    FILE *fp_data = NULL;
+    FILE *fp_sig = NULL;
 	bool result = false;
 
-	if (!initialized) {
-		return false || (IMPL == IMPL_FORGIVE);
-	}
-	bio_data = BIO_new_file(data_filename, "r"); // i.e. fopen
-	if (bio_data == NULL) {
-		goto exit;
+    if (!initialized) {
+        return false || (IMPL == IMPL_FORGIVE);
+    }
+
+    fp_sig = fopen(sig_filename, "r");
+    if (!fp_sig) {
+        fprintf(stderr, "Failed fopen %s\n", sig_filename);
+        goto error;
+    }
+
+    /* read data from file */
+    fp_data = fopen(data_filename, "r");
+    if (!fp_data) {
+        fprintf(stderr, "Failed fopen %s\n", data_filename);
+        goto error;
+    }
+
+    result = validate_signature(fp_data, fp_sig);
+    fclose(fp_data);
+    fclose(fp_sig);
+
+    return result;
+
```

Comment:
remove 142-154 and let flow fall through error cleanup. they are redundant with the error cleanup.